### PR TITLE
Perf and Memory Optimizations

### PR DESCRIPTION
**Pull request by vibhansa-msft**
Optimizations :
- Instead of returning handle ID to libfuse return the handle object address itself so that on read/write/flush/close calls we can save from map lookup to convert handl id to handle object
- Adding a new config which allows libfuse to directly read from locally cached file instead of forwarding request up in the pipeline
- various structures were not aligned properly leaving lot of padding in between, rearranged fields to minimize the memory usage
- lot of flags are used in handle/attr-cache/file-cache maps. Unified them all to use a bitmap instead of individual flags to conserve memory.
- sleep on file cache tests where we expect the file to take time to be deleted

### CODE REVIEW

**Review by gapra-msft — APPROVED** [review]

**Review by gapra-msft — APPROVED** [review]

**Review by souravgupta-msft — APPROVED** [review]

File: component/file_cache/file_cache_test.go [hunk 1]
```diff
@@ -995,8 +995,14 @@ func (suite *fileCacheTestSuite) TestRenameFileNotInCache() {
 	handle, _ := suite.fileCache.CreateFile(internal.CreateFileOptions{Name: src, Mode: 0777})
 	suite.fileCache.CloseFile(internal.CloseFileOptions{Handle: handle})
 
+	_, err := os.Stat(suite.cache_path + "/" + src)
+	for !os.IsNotExist(err) {
```
Comment (tasherif-msft) [inline component/file_cache/file_cache_test.go:999]: my only worry for this approach is if we run into a file cache bug where the file doesn't get deleted for one reason or another down the road. So maybe we should break after a given amount of time has passed and if we broke then the test fails

File: component/libfuse/libfuse2_handler.go [hunk 2]
```diff
@@ -440,8 +440,8 @@ func libfuse_create(path *C.char, mode C.mode_t, fi *C.fuse_file_info_t) C.int {
 		}
 	}
 
-	id := handlemap.Add(handle)
-	fi.fh = C.ulong(id)
+	handlemap.Add(handle)
+	fi.fh = C.ulong(uintptr(unsafe.Pointer(handle)))
```
Comment (tasherif-msft) [inline component/libfuse/libfuse2_handler.go:444]: was there a reason we didn't go with this approach from the start? why did we go with the handlemap approach first for the fuse layer?

**Review by tasherif-msft — APPROVED** [review]

File: component/libfuse/libfuse_handler.go [hunk 3]
```diff
@@ -185,6 +185,10 @@ func libfuse_init(conn *C.fuse_conn_info_t, cfg *C.fuse_config_t) (res unsafe.Po
 	conn.want |= C.FUSE_CAP_PARALLEL_DIROPS
 	conn.want |= C.FUSE_CAP_AUTO_INVAL_DATA
 	conn.want |= C.FUSE_CAP_READDIRPLUS
+	conn.want |= C.FUSE_CAP_ASYNC_READ
+
+	conn.max_background = 128
```
Comment (tasherif-msft) [inline component/libfuse/libfuse_handler.go:190]: could we add comments elaborating here please

**Comment by tasherif-msft** [comment]
We should add a test to confirm that libfuse doesn't hit file cache when the flag is set

File: component/libfuse/libfuse_handler_test_wrapper.go [hunk 4]
```diff
@@ -417,8 +411,7 @@ func testFsync(suite *libfuseTestSuite) {
 	openOptions := internal.OpenFileOptions{Name: name, Flags: flags, Mode: mode}
 	suite.mock.EXPECT().OpenFile(openOptions).Return(handle, nil)
 	libfuse_open(path, info)
-	handle.Path = name
-	handle.ID = handlemap.HandleID(info.fh)
+	handle = (*handlemap.Handle)(unsafe.Pointer(uintptr(info.fh)))
```
Comment (tasherif-msft) [inline component/libfuse/libfuse_handler_test_wrapper.go:414]: let's add a comment explaining what we're doing here, i.e, that we're converting the pointer to a handle and that it looks correct on the other component

File: component/file_cache/file_cache_test.go (hunk 1 above)
Comment (vibhansa-msft) [reply component/file_cache/file_cache_test.go:999]: Nice suggestion. Will see if we can break it after X retries.

Comment (vibhansa-msft) [reply component/file_cache/file_cache_test.go:999]: corrected on all instances of notExists check

File: component/libfuse/libfuse_handler.go (hunk 3 above)
Comment (vibhansa-msft) [reply component/libfuse/libfuse_handler.go:190]: added comments for all the flags

File: component/libfuse/libfuse_handler_test_wrapper.go (hunk 4 above)
Comment (vibhansa-msft) [reply component/libfuse/libfuse_handler_test_wrapper.go:414]: done

**Pull request merged**